lib/jsdom/living/helpers/shadow-dom.js

```diff
@@ -0,0 +1,252 @@
+"use strict";
+
+const NODE_TYPE = require("../node-type");
+const { domSymbolTree } = require("./internal-constants");
+
+const idlUtils = require("../generated/utils");
+
+// Valid host element for ShadowRoot.
+// Defined in: https://dom.spec.whatwg.org/#dom-element-attachshadow
+const VALID_HOST_ELEMENT_NAME = new Set([
+  "article",
+  "aside",
+  "blockquote",
+  "body",
+  "div",
+  "footer",
+  "h1",
+  "h2",
+  "h3",
+  "h4",
+  "h5",
+  "h6",
+  "header",
+  "main",
+  "nav",
+  "p",
+  "section",
+  "span"
+]);
+
+// TODO: Evaluate if it should be somewhere else.
+function isNode(nodeImpl) {
+  // Use an approximation by checking the presence of nodeType instead of instead of using the isImpl from
+  // "../generated/Node" to avoid introduction of circular dependencies.
+  return Boolean(nodeImpl && nodeImpl.nodeType);
+}
+
+function isShadowRoot(nodeImpl) {
+  // Use an approximation by checking the value of nodeType and presence of nodeType host instead of instead
+  // of using the isImpl from "../generated/ShadowRoot" to avoid introduction of circular dependencies.
+  return Boolean(nodeImpl && nodeImpl.nodeType === NODE_TYPE.DOCUMENT_FRAGMENT_NODE && nodeImpl.host);
+}
+
+// https://dom.spec.whatwg.org/#concept-slotable
+function isSlotable(nodeImpl) {
+  return nodeImpl && (nodeImpl.nodeType === NODE_TYPE.ELEMENT_NODE || nodeImpl.nodeType === NODE_TYPE.TEXT_NODE);
+}
+
+// https://dom.spec.whatwg.org/#concept-shadow-including-inclusive-ancestor
+function isShadowInclusiveAncestor(ancestor, node) {
+  while (isNode(node)) {
+    if (node === ancestor) {
+      return true;
+    }
+
+    if (isShadowRoot(node)) {
+      node = node.host;
+    } else {
+      node = domSymbolTree.parent(node);
+    }
+  }
+
+  return false;
+}
+
+// https://dom.spec.whatwg.org/#retarget
+function retarget(a, b) {
+  while (true) {
+    if (!isNode(a)) {
+      return a;
+    }
+
+    const aRoot = getRoot(a);
+    if (
+      !isShadowRoot(aRoot) ||
+      (isNode(b) && isShadowInclusiveAncestor(aRoot, b))
+    ) {
+      return a;
+    }
+
+    a = getRoot(a).host;
+  }
+}
+
+// https://dom.spec.whatwg.org/#get-the-parent
+function getParent(eventTarget, event) {
```

Comment:
I would rather implement this as private methods on ShadowRoot, Document, and Node, that are called polymorphically. That would be closer to the spec.

Reply:
I went down this route originally. However, adding `getParent` private method requires the invoker to invoke it defensively.

In the case where you want to climb the parent chain, invoking getParent on `Document` return the global object but invoking getParent on the global object  throws since the global object doesn't have the private method. So you need to guard the getParent invocation with an if statement. The `getParent` method doesn't suffer from the same issue.

What about an intermediary approach where `getParent` is a function invoking the private method if present, otherwise returns `null`. This approach would be closer to the spec without requiring the invoker to be defensive.

```js
funtion getParent(nodeImpl, eventImpl) {
	return "_getParent" in nodeImpl ? nodImpl.getParent(eventImpl) : null;
}

class DocumentImpl {
	// ...

	_getParent(eventImpl) {
		// The getParent logic on document
	}
}
```

Reply:
I tried to move the `_getParent` to the window object, however, I ran in issues since the window impl is not like the standard ones. I don't know yet where do the issues come from.